packages/flame/lib/src/components/mixins/has_game_ref.dart

```diff
@@ -1,41 +1,54 @@
-import 'package:flame/components.dart';
-import 'package:flame/game.dart';
-import 'package:flutter/material.dart';
+import 'package:flame/src/components/core/component.dart';
+import 'package:flame/src/game/game.dart';
+import 'package:flame/src/game/mixins/single_game_instance.dart';
+import 'package:meta/meta.dart';
 
-mixin HasGameRef<T extends FlameGame> on Component {
-  T? _gameRef;
+/// [HasGameRef] mixin provides property [game] (or [gameRef]), which is the
+/// cached accessor for the top-level game instance.
+///
+/// The type [T] on the mixin is the type of your game class. This type will be
+/// the type of the [game] reference, and the mixin will check at runtime that
+/// the actual type matches the expectation.
+mixin HasGameRef<T extends Game> on Component {
```

Comment:
This is a breaking change changing this from `FlameGame` to `Game`, now I understand why you had to add the explicit generics on the other ones. I understand this is to make this usable by `BareFlameGame`, but is there no other way to do this? Maybe having `BareFlameGame` actually extending `FlameGame` and just throw an `Unimplemented exception` if the camera is accessed?

Reply:
I would characterize this as "barely breaking", because in practical usage one would always derive from `FlameGame`:
```dart
class MyGame extends FlameGame with HasTappableComponents, ... {
}
```
and then when using `HasGameRef`, you'd specify your game class:
```dart
class SomeComponent extends PositionComponent with HasGameRef<MyGame> {
}
```

Really, it is only in very corner cases that you'd want to build the game using the plain "FlameGame" class -- such as in simple examples or tests.

Reply:
It's very normal to not specify anything there, if you for example want the size or camera etc that you can access directly from the `FlameGame`, so I'd definitely not call this a minor breaking change.

Reply:
Ok, so should I add the migration instructions? Or revert to the way it was before, with a separate mixin?

Reply:
But do we really need it? Can't the `BareFlameGame` just extend `FlameGame` in the manner I described above instead?

Reply:
That would mean it's not "bare" anymore. We could, in theory, make `FlameGame` derived from `BareFlameGame`, but not the the other way around. 

For example, making `BareFlameGame` derived from `FlameGame` would make it impossible to declare a `camera` property which is incompatible with the FlameGame's camera.

Reply:
What if we keep `HasGameReference` as it was originally (separate from `HasGameRef`), but inside the `experimental` folder?

Reply:
Yeah, we'd just restrict that field and return an unimplemented exception instead, the game would still be bare since it wouldn't use the camera. People can just declare `cameraComponent` if they want to store their camera in a field. I think that is much less messy than putting in this breaking change that breaks lots of tutorials. And I don't think it's a good idea to have both, it's better to wait with those changes until V2.

Reply:
The whole idea of creating new clean implementations is so that we can leave behind the cruft that accumulated in the old classes. We cannot undermine architectural integrity of the foundational classes in order to get a minor temporary advantage like that.

Reply:
On the other hand, it is not that bad if they choose the wrong one here if the only difference is the generics and it won't break current implementations. Alright, I'm onboard with creating `HasGameReference` too in experimental then, and then just have them identical (so the current changes doesn't have to be reverted) except for the generics part.

Reply:
@st-pasha did you see my last comment here?

Reply:
Yep, just no time right now to work on changes

Reply:
Done